### PR TITLE
Connect the revenue report to the API

### PR DESCRIPTION
**Pull request by emdashcodes**
This PR is the first step in properly connecting the revenue report up to the API.

The report is not 100% done. Here are a few notes/todos/issues.

* Zero-filled data. After the meeting today, we decided this will be done server side. In in the meantime, if you have gap days, they will not show up on the report. That means some report filters might return an 'empty' response (i.e. if you have 'today' selected with no orders) right now. This can also cause an issue with the 'Previous Period' total on the chart (see screenshot below), because we pull out the matching secondary data point based on the primary dates, and if some are missing from the response, the secondary points don't get added to the chart.
* Different intervals. Right now, you can only view data points as days. More handling needs to be written to support other intervals (hours, weeks, etc).
* No loading indicators or empty state. I have some super temporary unstyled messages implemented at the moment. We still need a full page loading indicator and a message for empty/error states. See #302.
* The data points seem to be off by one day (i.e. the axis shows 29, but if you hover over the point it is for the 30th). This seems to be happening on the hardcoded dashboard example as well. I will log an issue. cc @greenafrican 
* The utilities and code I've written for this report should be useable in other reports. I've left the code in revenue for now while trying to figure everything out, but I expect this code to be moved out so it can be reused by other reports.

Otherwise, you are able to filter the report by date, paginate and sort the table, and view summary numbers.

A note on the API requests: For requests that have results over multiple API pages (i.e. you are viewing a large date span with over 100 results), we need to make requests for all pages so we can properly fill out the graph. I created a utility to help with this.

screen shot 2018-08-22 at 1 16 33 pm

screen shot 2018-08-22 at 1 17 12 pm

To Test:
* run `npm test` and make sure all tests pass
* Before testing on your site, you may want to generate a number of orders. You can use [wc-smooth-generator](https://github.com/woocommerce/wc-smooth-generator) to do this. See the [README](https://github.com/woocommerce/wc-smooth-generator/blob/master/README.md#orders) for an example.
* Pull the latest WooCommerce v3 API from `feature/rest-api-v3-reports` (you might need to deactivate `wc-api-dev`).
* You need to deactivate and reactivate WooCommerce to get the correct summary MySQL tables. There is also a tool under system settings to rebuild the summary tables.
* Visit the revenue report.
* Test reordering columns, paging results, etc.

### CODE REVIEW

**Comment by timmyc** [comment]
I'm getting some errors from the API while trying to test this branch - I am on the most current version of `feature/rest-api-v3-reports`

FWIW this is a pretty "clean" install of WooCommerce, all orders and products ( except for one order ) were created using `smooth-generator`

/cc @peterfabian 

```
<br />
<b>Notice</b>:  Trying to get property 'totals' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>66</b><br />
<br />
<b>Warning</b>:  get_object_vars() expects parameter 1 to be object, null given in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>66</b><br />
<br />
<b>Notice</b>:  Trying to get property 'intervals' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>70</b><br />
<br />
<b>Warning</b>:  Invalid argument supplied for foreach() in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>70</b><br />
<br />
<b>Notice</b>:  Trying to get property 'total' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>76</b><br />
<br />
<b>Notice</b>:  Trying to get property 'pages' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>77</b><br />
<br />
<b>Notice</b>:  Trying to get property 'page_no' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>79</b><br />
<br />
<b>Notice</b>:  Trying to get property 'pages' of non-object in <b>/srv/www/example/public_html/wp-content/plugins/woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php</b> on line <b>80</b><br />
{"totals":null,"intervals":[]}
```

**Review by timmyc — COMMENTED** [review]
Super awesome to see this PR! Left some comments, and will try to see why I'm getting errors with the API requests.

File: client/analytics/report/revenue/index.js [hunk 1]
```diff
@@ -269,42 +261,107 @@ class RevenueReport extends Component {
...
 	}
 
-	render() {
-		const { path, query } = this.props;
-		const { totals = {}, intervals = [] } = this.state.stats;
-		const summary = this.getSummaryContent( totals ) || [];
-		const rows = this.getRowsContent( intervals ) || [];
+	renderTable() {
+		const { primaryData, query } = this.props;
+		const intervals = primaryData.data.intervals;
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/analytics/report/revenue/index.js]: We check to see if `primaryData` is empty/doesn't contain in the `render()` function and will return a message - so `renderTable()` never gets called.

File: client/lib/date/index.js [hunk 2]
```diff
@@ -271,18 +271,47 @@ export const getCurrentDates = query => {
...
 };
 
+/**
+ * Calculates the date difference between two dates. Used in calculating a matching date for previous period.
+ *
+ * @param {String} date - Date to compare
+ * @param {String} date2 - Seconary date to compare
+ * @return {Int}  - Difference in days.
+ */
+export const getDateDifference = ( date, date2 ) => {
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/lib/date/index.js]: I'll still rename it for now. I might still just need the difference in days for the other intervals.

File: client/components/pagination/index.js [hunk 3]
```diff
@@ -167,7 +169,10 @@ class Pagination extends Component {
 
 		if ( this.pageCount <= 1 ) {
 			return (
-				( total > 25 && <div className={ classes }>{ this.renderPerPagePicker() }</div> ) || null
+				( total > PER_PAGE_OPTIONS[ 0 ] && (
```
Comment (timmyc) [inline client/components/pagination/index.js]: 👍 thanks for this update

**Comment by justinshreve** [comment]
@timmyc I've handled all your feedback above.

I've also updated this branch to use the new generic stats implementation (and added any necessary functionality), and dropped the existing revenue files.

**Review by timmyc — COMMENTED** [review]
updates are looking good, thanks! i'll try testing this against another site to see if i can get past my API errors and give this a final 👍

**Comment by timmyc** [comment]
So I setup a brand new test site using the `feature/rest-api-v3-reports` branch of woocommerce, and this branch of `wc-admin` running on a new vvv instance. With that setup I'm still getting the error I noted above returned from the endpoint... I also manually ran the "Rebuild reports data" and waiting fifteen minutes.

It appears `$report_data     = $reports_revenue->get_data();` was returning nothing, and that is what resulted in all the errors in the API. This diff adds in some non-elegant testing for various things and actually allows the endpoint to return an empty response without errors:

https://gist.github.com/timmyc/e41748741d2dbbd9e003f85827f5ead6

Not sure if I'm missing a step with setting up some database tables or something else

**Comment by peterfabian** [comment]
@timmyc Tried on a clean install of WP & WC 3.4.4, created 2 products and 2 order manually, then added 20 products and 100 orders using smooth generator, checked out `feature/rest-api-v3-reports`, reactivated the plugin, clicked on `Rebuild reports` and it generated the data into the db table `wp_wc_order_stats` and then `wc/v3/reports/revenue/stats/` worked ok.

Based on the error you reported, it seems like there are basically 2 possible explanations: 
- result from db being invalid 
- page requested is < 1 or > total number of pages

Maybe you can check how the table `wp_wc_order_stats` looks in your installation--is there any data in there? Any errors in the logs that might relate to generating data for it?

**Comment by ryelle** [comment]
I was having the same problem as @timmyc - no data returned from the API - but in my case I really didn't have data for the default range. If I change the date query to something I know I have data for, like "Quarter to date", it does render the report.

On the other hand, I'm not sure the default query returns any results. It looks like the page defaults to fetching for the last day by default. but the endpoint it requests is for "before 8/27 and after 8/27", which I think would always return no orders?

`wp-json/wc/v3/reports/revenue/stats?order=asc&interval=day&per_page=100&after=2018-08-27T00%3A00%3A00%2B00%3A00&before=2018-08-27T00%3A00%3A00%2B00%3A00`

In any case, it seems like the end date is not correct in "Week to date" either. I created an order today (8/27), and I see it in `wp_wc_order_stats`, but I get an error on the "Week to date" report, even though it reads `Aug 26 - 27, 2018`. Again, it asks for `before=2018-08-27…`.

**Comment by ryelle** [comment]
Something seems off with the graph here, according to the SummaryNumber the previous period did have some shipping data, so I think this is being incorrectly zeroed:

screen shot 2018-08-27 at 11 23 55 am

**Comment by justinshreve** [comment]
Thanks for noting the query. The `before` date should actually be for the end of the day (`23:59:59`) instead of `00:00`, otherwise it will never include that day in the orders. I pushed a fix.

In terms of the chart, this looks due to lack of "zero-filling" data I mentioned above. The way the chart is rendered is we take all of the current data points, and then calculate their matching secondary data point, and then attach them both to the same axis. That means a secondary data point will only show up right now if there was also a data point for the 'primary'. (i.e. you have data for `Jan 1st` but not `April 1st`, the Jan data point is never output on the graph).

This will be fixed on the API side because all days will be returned in the results, so it didn't seem worth trying to fix with reverse calculation in the meantime.

**Comment by justinshreve** [comment]
@timmyc and I talked last week and we decided to get this merged so we can iterate on it in follow up PRs. 

I'll get this rebased and merged.

The issues that still need handled:

* Zero-filling. This is happening on the API side. cc @peterfabian 
* Intervals. Only day is supported right now.
* UI States. Loading indicators and error state.
* D3 Chart Issues. Flagged here: https://github.com/woocommerce/wc-admin/issues/320.

**Pull request merged**